### PR TITLE
Add deploy type for lambda functions

### PR DESCRIPTION
**Pull request by grvcoelho**
## Description

Add a new form of Deploy: Lambda. It performs the following actions:

1. Take an array of files and folders (`deploy.package`) and zips it.
2. Sends the zip to AWS S3 to a specified bucket (`deploy.s3_bucket`) and to an S3 key that is the sha of the commit. E.g. when deploying commit `abcd123` to the bucket `my-bucket`, the deploy will create a key on AWS S3 named `my-bucket/abcd123/package.zip`. The sha of the commit comes from the SCM step.
3. Take an array of lambda functions (`deploy.functions`) and build the name of the functions with the template `${environment}-${projectName}-${functionName}`. E.g. given the function name `hello` being deployed on the `acme` project on the `live` environment, the final name of the function will be `live-acme-hello`.
4. Updates the functions code on AWS Lambda.

### Usage

On deployer.yml:

```yml
deploy:
  type: lambda
  project_name: acme
  s3_bucket: acme-deploy-bucket
  package:
    - dist
    - node_modules
  functions:
    - fizz
    - hello
```

### CODE REVIEW

File: deploy/lambda/lambda.go [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+	uploader := s3manager.NewUploader(sess)
+	file, err := os.Open(filename)
+
+	if err != nil {
+		return errors.New("could not open zip file")
+	}
+
+	defer file.Close()
+
+	_, err = uploader.Upload(&s3manager.UploadInput{
```
Comment (wilkmaia) [inline deploy/lambda/lambda.go]: You should check `err` for errors :)

File: deploy/lambda/lambda.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+	select {
+	case <-doneChannel:
+		return nil
+	case err := <-errChannel:
+		return err
+	}
+}
+
+func (l *Lambda) Deploy(ctx pipeline.Context) error {
+	env := ctx["Environment"].(map[string]interface{})["name"].(string)
```
Comment (wilkmaia) [inline deploy/lambda/lambda.go]: A lot of error ignoring going on here.

Sample error handling by @derekstavis 
```go
env, ok := ctx["Environment"]

if !ok {
  // error
}

envMap, ok := env.(map[string]interface{})

if !ok {
  // error
}

name, ok := envMap["name"]

if !ok {
  // error
}

nameString, ok := name.(string)

if !ok {
  // error
}
```

Comment (wilkmaia) [inline deploy/lambda/lambda.go]: If something goes wrong there and you're not aware, your code may explode somewhere ahead and make it a little bit harder to debug :)

File: deploy/lambda/lambda.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+	ProjectName string   `mapstructure:"project_name"`
+	Functions   []string `mapstructure:"functions"`
+	Package     []string `mapstructure:"package"`
+}
+
+type Lambda struct {
+	Options *Options
+}
+
+func (l *Lambda) zipPackage(path, filename string) (string, error) {
```
Comment (greenboxal) [inline deploy/lambda/lambda.go]: O ZIP em si poderia ser outro step.

**Comment by grvcoelho** [comment]
@greenboxal: The `npm install` part will be performed in a Rockerfile (using the rocker build step). There's a lot of advantages in this approach: we can `npm install` inside a linux machine (this will help us avoid node binary errors that are built on MacOs and executed on Linux inside Lambdas), and we generally have more control over the build part.

File: deploy/lambda/lambda.go (hunk 2 above)
Comment (grvcoelho) [inline deploy/lambda/lambda.go]: Thanks :+1: I'll fix it 🙂

File: deploy/lambda/lambda.go (hunk 3 above)
Comment (derekstavis) [inline deploy/lambda/lambda.go]: Concordo, kk

File: deploy/lambda/lambda.go [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+	}
+
+	if err := archiver.Zip.Make(packagePath, packages); err != nil {
+		return "", err
+	}
+
+	return packagePath, nil
+}
+
+func (l *Lambda) uploadZipToS3(filename, S3Key string) error {
```
Comment (derekstavis) [inline deploy/lambda/lambda.go]: O upload pro S3 poderia ser outro step também!

File: deploy/lambda/lambda.go (hunk 3 above)
Comment (grvcoelho) [inline deploy/lambda/lambda.go]: Will discuss on #4 and do it in a later PR

**Pull request merged**